### ISSUE TITLE
MovingAverageEarlyStopping

### ISSUE THREAD

**Issue by nikolaosmparoutis** [keras-team/keras#18210]
If you open a GitHub issue, here is our policy:

It must be a bug, a feature request, or a significant problem with the documentation (for small docs fixes please send a PR instead).
The form below must be filled out.

**Here's why we have that policy:**.

Keras developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

**System information**.

TensorFlow version (you are using): **v2.12.0**
Are you willing to contribute it (Yes/No) : **Yes**

**Describe the feature and the current behavior/state**.
The requested feature is a custom early stopping callback for TensorFlow Keras models called `EarlyStoppingMovingAverage`. This callback extends the built-in `EarlyStopping` class and adds functionality to monitor the moving average (either simple moving average or exponential moving average) of a specified metric instead of the raw metric values.
The current behavior of the built-in `EarlyStopping` callback is to monitor a single metric (e.g., validation loss) and stop training when the metric does not improve for a specified number of consecutive epochs (patience). However, this approach can be sensitive to noisy or fluctuating metric values, leading to premature stopping.
The custom `EarlyStoppingMovingAverage` callback addresses this issue by monitoring the moving average of the metric, providing a more stable and reliable stopping criterion. This feature is needed to prevent premature stopping and improve model performance, especially in cases where the monitored metric values are noisy or fluctuate significantly during training. The use-case for this feature includes training complex models on challenging datasets where the raw metric values may not provide a clear indication of the model's performance trend.

**Will this change the current api? How?** This custom early stopping feature, `EarlyStoppingMovingAverage`, will not change the current TensorFlow Keras API directly, as it is a user-defined extension of the built-in `EarlyStopping` callback. It built to provide an additional option for users who want to implement early stopping based on the moving average of a monitored metric.
Like we did instead: 
`from tensorflow.keras.callbacks import EarlyStopping`
Now also can do:
`from tensorflow.keras.callbacks import EarlyStoppingMovingAverage`  

**Who will benefit from this feature?**
This custom early stopping feature will benefit users working with TensorFlow Keras models, particularly when the monitored metric values are noisy or fluctuate significantly during training. It can help prevent premature stopping and lead to better model performance, making it useful for researchers, data scientists, and machine learning engineers.
When datasets have many flactuations during training and validation loss does many spikes the common early stoping may stop too early the training if finds a big spike too early. Also if the training goes too far in training epochs when increase the patience could be better to stop in an earlier global or to sub-global minimum point.

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**

- Do you want to contribute a PR? (yes/no): **yes**
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing):

It extends the built-in `EarlyStopping` class from `tensorflow.keras.callbacks` and adds functionality to monitor the moving average of a specified metric (either simple moving average or exponential moving average) instead of the raw metric values.
The `EarlyStoppingMovingAverage` class takes additional parameters like `window_size` and `moving_average_type` to control the moving average calculation. The moving average is calculated in the `_calculate_sma` and `_calculate_ema` methods, and the improvement is checked in the `_is_improvement_sma` and `_is_improvement_ema` methods.
The custom early stopping callback is designed to be more robust against noisy metric values and can help prevent premature stopping in cases where the raw metric values fluctuate significantly. By monitoring the moving average of the metric, the callback can better identify the general trend and make more informed decisions about when to stop training.

**Comment by tilakrayal** [comment]
@nikolaosmparoutis,
Could you please try to provide any Use-case for the mentioned support which helps us to analyse the issue in an effective way. Thank you!

**Comment by nikolaosmparoutis** [comment]
Yes, i can provide 2 example figures from val_losses  to see what i mean.Then i will wait for your logical feedback and i will study it further.
Here are 4 figs: 2 with 'val_losses' two with 'val_acc'.
Using the standard early stopping method with a 'min' function may result in premature termination if a small and reasonable patience parameter (2-5) is applied. This does not necessarily indicate that the model is fully trained. If the patience parameter is significantly increased (>10), the training might stop too late due to the occurrence of a minimum spike, which could lead to overfitting. However, this can be mitigated by using an exponential moving average, which allows you to adjust the stopping point based on the weight you assign to recent versus past data. If you believe recent data is more important, the trend of the metric will stop later, taking into account the current data. Conversely, if you believe the trend should be based more on past data, the stopping point will be more resistant to current fluctuations and will stop if a more significant slope is observed.
For example:
1. Small patience parameter (e.g., 2): After each epoch, we check the validation loss. If the validation loss doesn't decrease for 2 consecutive epochs, we stop the training. This might lead to premature stopping. For example, it's possible that after 2 epochs with no decrease, the 3rd epoch could have shown improvement. But because our patience was set to 2, we stopped early and potentially missed out on a better model.

2. Large patience parameter (e.g., 15): we wait for 15 epochs without improvement before stopping. This might lead to overfitting. For instance, we might see a temporary increase in validation loss due to some noise in the data, but because our patience is high, we keep training. This could result in a model that's too complex and performs well on the training data but poorly on unseen data.
  
![W B Chart 14_6_2023, 2_49_10 μ μ](https://github.com/keras-team/keras/assets/7507597/fcbae1ce-5db7-479a-a2c6-6f3824b9b35e)
 
![W B Chart 14_6_2023, 2_48_12 μ μ](https://github.com/keras-team/keras/assets/7507597/60594dbe-7338-405e-b19b-d04ae75452a0)
![W B Chart 14_6_2023, 2_48_53 μ μ](https://github.com/keras-team/keras/assets/7507597/2bae1fc1-9ea7-47ce-a380-a02731c35668)




![W B Chart 14_6_2023, 2_49_18 μ μ](https://github.com/keras-team/keras/assets/7507597/47363968-ad56-496d-9c80-4b652164e9df)

**Comment by hertschuh** [comment]
Hi Nikolaos,

Thank you for the detailed feature request entry.

One approach to solving this, which we think is the easiest, would be the following:
* Create a `MovingAverageMetric` class and callback. What it does is implement `on_epoch_end`, take the value of the specific metric it's supposed to average, do the math to average it in the way you want, and then update its value.
* Use the normal `EarlyStopping` callback, but configure it to use the specific `MovingAverageMetric` instead of the raw metric.

This should cover this use case with very few lines of code.

Thanks,
Fabien
